src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs

```diff
@@ -83,7 +84,9 @@ public static partial class JsonSerializer
 
                     // An object or another enumerator requires a new stack frame.
                     state.Push(elementClassInfo, value);
-                    if (options?.DictionaryKeyPolicy != null && state.Current.ExtensionDataStatus != ExtensionDataWriteStatus.Writing)
+                    var extensionDataAttribute = jsonPropertyInfo.PropertyInfo?.GetCustomAttribute(typeof(JsonExtensionDataAttribute));
```

Comment:
Is there a specific reason to check the Attribute? 
`GetCustomAttribute` uses reflection which is very expensive. I was thinking that the sole condition of `state.Current.ExtensionDataStatus != ExtensionDataWriteStatus.Writing` was enough. did you find any issues while just depending on the condition?

Reply:
Yes, it is heavy operation, but we need this check for passing test **SerializationWithJsonExtensionDataAttribute_IgoneDictionaryKeyPolicy**, I'm debugging this test (with extensionDataAttribute) for non-value class types, they were affected by DictionaryKeyPolicy, _state.Current.ExtensionDataStatus_ had **NotStarted** value.
Without checking attribute, I've got next:
<pre>
"{""KeyInt"":1000,""KeyString"":""text"",""KeyBool"":true,""keyObject"":{},""keyList"":[],""keyDictionary"":{}}"</pre>

Reply:
I see the problem.
The reason why you get `NotStarted` is because, as of now, you are checking *after* one of the *values* of the `JsonExtensionData` dictionary is pushed into the stack in order to be written on the next iteration. The *state* of said value does not carry the parent dictionary's `ExtensionDataStatus` and therefore you see it as `NotStarted`.

To check against the proper `ExtensionDataStatus` try to move the condition before calling `state.Push`.